File: backend/src/modules/hr/tickets.ts
import { Router, type Request } from 'express'
import type pg from 'pg'
import { z } from 'zod'
import { query, tx } from '../../db/pool'
import { buildUpdate } from '../../db/sql'
import { audit } from '../../lib/audit'
import { badRequest, forbidden, notFound, parse } from '../../lib/http'
import { isExec, isLeader } from '../../lib/roles'
import { auth, type AuthContext } from '../../middleware/auth'
import { TICKET_SELECT, toTicket } from '../workspace/repository'
import { isoDate } from './helpers'

export const ticketsRouter = Router()

const STATUSES = ['Backlog', 'Todo', 'In Progress', 'In Review', 'Done', 'Canceled'] as const
const PRIORITIES = ['Urgent', 'High', 'Medium', 'Low', 'None'] as const

const label = z.string().trim().toLowerCase().min(1).max(32)
const labels = z.array(label).max(10).transform((l) => [...new Set(l)])

/** Params for TICKET_SELECT: workspace, viewer, and whether the viewer may see private (HR) tickets. */
const scope = (me: AuthContext) => [me.workspaceId, me.employeeId, isExec(me.role)]

async function loadTicket(db: pg.PoolClient, me: AuthContext, id: string, lock = false) {
  if (lock) await db.query('SELECT 1 FROM tickets WHERE id = $1 AND workspace_id = $2 FOR UPDATE', [id, me.workspaceId])
  const { rows } = await db.query(`${TICKET_SELECT} AND t.id = $4`, [...scope(me), id])
  if (!rows[0]) throw notFound('Ticket')
  return rows[0] as Record<string, any>
}

async function lockTeam(db: pg.PoolClient, workspaceId: string, teamId: string) {
  const { rows } = await db.query<{ id: string; key: string; name: string }>('SELECT id, key, name FROM ticket_teams WHERE id = $1 AND workspace_id = $2 FOR UPDATE', [teamId, workspaceId])
  if (!rows[0]) throw badRequest('Unknown team')
  // Holding the team row lock serialises number allocation for this team.
  const { rows: seq } = await db.query<{ next: number }>('SELECT COALESCE(MAX(number), 0) + 1 AS next FROM tickets WHERE team_id = $1', [teamId])
  return { team: rows[0], number: Number(seq[0]!.next) }
}

async function assertEmployee(db: pg.PoolClient, workspaceId: string, employeeId: string) {
  const { rowCount } = await db.query("SELECT 1 FROM employees WHERE id = $1 AND workspace_id = $2 AND status <> 'Exited'", [employeeId, workspaceId])
  if (!rowCount) throw badRequest('Assignee must be an active member of this workspace')
}

async function notifyAssignee(db: pg.PoolClient, req: Request, assigneeId: string, ticket: { id: string; identifier: string; title: string }) {
  const me = auth(req)
  if (assigneeId === me.employeeId) return
  await db.query(
    `INSERT INTO notifications (workspace_id, recipient_id, type, title, body, href)
     SELECT $1, $2, 'system', $3, e.name || ' assigned this ticket to you.', $4 FROM employees e WHERE e.id = $5`,
    [me.workspaceId, assigneeId, `${ticket.identifier} · ${ticket.title}`, `/app/tickets?id=${encodeURIComponent(ticket.id)}`, me.employeeId],
  )
}

// ── List & read ─────────────────────────────────────────────────────
ticketsRouter.get('/tickets', async (req, res) => {
  const me = auth(req)
  const f = parse(
    z.object({
      team: z.string().trim().max(64).optional(),
      status: z.string().optional(),
      assignee: z.literal('me').optional(),
      reporter: z.literal('me').optional(),
      q: z.string().trim().max(120).optional(),
    }),
    req.query,
  )
  const statuses = f.status ? parse(z.array(z.enum(STATUSES)), f.status.split(',')) : null
  const params: unknown[] = [...scope(me)]
  const where: string[] = []
  if (f.team) {
    params.push(f.team)
    where.push(`(t.team_id = $${params.length} OR tm.key = upper($${params.length}))`)
  }
  if (statuses) {
    params.push(statuses)
    where.push(`t.status = ANY ($${params.length}::text[])`)
  }
  if (f.assignee) where.push('t.assignee_id = $2')
  if (f.reporter) where.push('t.reporter_id = $2')
  if (f.q) {
    params.push(`%${f.q.replace(/[\\%_]/g, (c) => `\\${c}`)}%`)
    where.push(`(t.title ILIKE $${params.length} OR t.identifier ILIKE $${params.length} OR t.description ILIKE $${params.length})`)
  }
  const rows = await query(`${TICKET_SELECT}${where.map((w) => ` AND ${w}`).join('')} ORDER BY t.updated_at DESC LIMIT 500`, params)
  res.json(rows.map(toTicket))
})

ticketsRouter.get('/tickets/:id', async (req, res) => {
  const me = auth(req)
  const [row] = await query(`${TICKET_SELECT} AND (t.id = $4 OR t.identifier = upper($4))`, [...scope(me), req.params.id])
  if (!row) throw notFound('Ticket')
  res.json(toTicket(row))
})

// ── Create ──────────────────────────────────────────────────────────
const createSchema = z.object({
  teamId: z.string().min(1),
  title: z.string().trim().min(3, 'Give the ticket a short title').max(200),
  description: z.string().trim().max(10_000).default(''),
  status: z.enum(STATUSES).default('Todo'),
  priority: z.enum(PRIORITIES).default('None'),
  assigneeId: z.string().min(1).nullish(),
  labels: labels.default([]),
  dueDate: z.string().regex(isoDate).nullish(),
})

ticketsRouter.post('/tickets', async (req, res) => {
  const me = auth(req)
  const b = parse(createSchema, req.body)
  const ticket = await tx(async (db) => {
    const { team, number } = await lockTeam(db, me.workspaceId, b.teamId)
    if (b.assigneeId) await assertEmployee(db, me.workspaceId, b.assigneeId)
    const identifier = `${team.key}-${number}`
    const { rows } = await db.query<{ id: string }>(
      `INSERT INTO tickets (workspace_id, team_id, number, identifier, title, description, status, priority, reporter_id, assignee_id, labels, due_date, completed_at)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12, CASE WHEN $7 = 'Done' THEN now() END) RETURNING id`,
      [me.workspaceId, team.id, number, identifier, b.title, b.description, b.status, b.priority, me.employeeId, b.assigneeId ?? null, b.labels, b.dueDate ?? null],
    )
    const id = rows[0]!.id
    if (b.assigneeId) await notifyAssignee(db, req, b.assigneeId, { id, identifier, title: b.title })
    await audit(req, 'ticket.created', 'ticket', id, { identifier, team: team.key, priority: b.priority }, db)
    return toTicket(await loadTicket(db, me, id))
  })
  res.status(201).json(ticket)
})

// ── Update ──────────────────────────────────────────────────────────
const patchSchema = z
  .object({
    title: z.string().trim().min(3).max(200),
    description: z.string().trim().max(10_000),
    status: z.enum(STATUSES),
    priority: z.enum(PRIORITIES),
    assigneeId: z.string().min(1).nullable(),
    labels,
    dueDate: z.string().regex(isoDate).nullable(),
    teamId: z.string().min(1),
  })
  .partial()
  .strict()

ticketsRouter.patch('/tickets/:id', async (req, res) => {
  const me = auth(req)
  const b = parse(patchSchema, req.body)
  if (!Object.keys(b).length) throw badRequest('Nothing to update')

  const ticket = await tx(async (db) => {
    const cur = await loadTicket(db, me, req.params.id, true)
    const involved = cur.reporter_id === me.employeeId || cur.assignee_id === me.employeeId
    // Anyone may pick up an unassigned ticket for themselves (Linear's "assign to me").
    const selfAssignOnly = Object.keys(b).length === 1 && b.assigneeId === me.employeeId && !cur.assignee_id
    if (!involved && !isLeader(me.role) && !selfAssignOnly) throw forbidden('Only the reporter, the assignee, managers and HR can edit this ticket')

    if (b.assigneeId) await assertEmployee(db, me.workspaceId, b.assigneeId)

    const patch: Record<string, unknown> = { ...b }
    let identifier = cur.identifier as string
    if (b.teamId && b.teamId !== cur.team_id) {
      const { team, number } = await lockTeam(db, me.workspaceId, b.teamId)
      identifier = `${team.key}-${number}`
      Object.assign(patch, { number, identifier })
    } else delete patch.teamId
    if (b.status && b.status !== cur.status) patch.completedAt = b.status === 'Done' ? new Date() : null

    const set = buildUpdate(patch, ['title', 'description', 'status', 'priority', 'assignee_id', 'labels', 'due_date', 'team_id', 'number', 'identifier', 'completed_at'], 3)
    if (set) {
      await db.query(
        `UPDATE tickets SET ${set.sql}, updated_at = now() WHERE id = $1 AND workspace_id = $2`,
        [cur.id, me.workspaceId, ...set.params],
      )
    }

    if (b.status && b.status !== cur.status) {
      await audit(req, 'ticket.status_changed', 'ticket', cur.id, { identifier, from: cur.status, to: b.status }, db)
    }
    if (b.teamId && b.teamId !== cur.team_id) {
      await audit(req, 'ticket.moved', 'ticket', cur.id, { from: cur.identifier, to: identifier }, db)
    }
    if (b.assigneeId && b.assigneeId !== cur.assignee_id) {
      await notifyAssignee(db, req, b.assigneeId, { id: cur.id, identifier, title: b.title ?? cur.title })
      await audit(req, 'ticket.assigned', 'ticket', cur.id, { identifier, assigneeId: b.assigneeId }, db)
    }
    return toTicket(await loadTicket(db, me, cur.id))
  })
  res.json(ticket)
})

// ── Comments ────────────────────────────────────────────────────────
ticketsRouter.post('/tickets/:id/comments', async (req, res) => {
  const me = auth(req)
  const { body } = parse(z.object({ body: z.string().trim().min(1, 'Write a comment first').max(5000) }), req.body)
  const comment = await tx(async (db) => {
    const cur = await loadTicket(db, me, req.params.id)
    const { rows } = await db.query(
      'INSERT INTO ticket_comments (workspace_id, ticket_id, author_id, body) VALUES ($1, $2, $3, $4) RETURNING id, author_id, body, created_at',
      [me.workspaceId, cur.id, me.employeeId, body],
    )
    await db.query('UPDATE tickets SET updated_at = now() WHERE id = $1', [cur.id])
    // Let the other people on the ticket know (one statement, no fan-out).
    await db.query(
      `INSERT INTO notifications (workspace_id, recipient_id, type, title, body, href)
       SELECT $1, r.id, 'system', $2, left($3, 180), $4
         FROM (SELECT DISTINCT unnest(ARRAY[$5::text, $6::text]) AS id) r
        WHERE r.id IS NOT NULL AND r.id <> $7`,
      [me.workspaceId, `New comment on ${cur.identifier}`, body, `/app/tickets?id=${encodeURIComponent(cur.id)}`, cur.reporter_id, cur.assignee_id, me.employeeId],
    )
    return rows[0]!
  })
  res.status(201).json({ id: comment.id, authorId: comment.author_id, body: comment.body, createdAt: new Date(comment.created_at).toISOString() })
})

File: backend/api/routers/tickets.py
"""Linear-style tickets: teams, per-team identifiers, comments and assignee notifications."""

import datetime as dt
import re
from typing import Annotated, Any, Literal
from urllib.parse import quote

import psycopg
from fastapi import APIRouter, Body, Depends, Request
from fastapi.responses import JSONResponse
from pydantic import AfterValidator, BaseModel, ConfigDict, Field, RootModel, StringConstraints
from pydantic_core import PydanticCustomError

from ..audit import audit
from ..db import build_update, iso, query, query_one, tx
from ..errors import bad_request, forbidden, not_found, parse
from ..repository import TICKET_SELECT, to_ticket
from ..roles import is_exec, is_leader
from ..security import AuthContext, require_auth

router = APIRouter()

ISO_DATE = r"^\d{4}-\d{2}-\d{2}$"
Status = Literal["Backlog", "Todo", "In Progress", "In Review", "Done", "Canceled"]
Priority = Literal["Urgent", "High", "Medium", "Low", "None"]

Label = Annotated[str, StringConstraints(strip_whitespace=True, to_lower=True, min_length=1, max_length=32)]
Labels = Annotated[list[Label], Field(max_length=10), AfterValidator(lambda l: list(dict.fromkeys(l)))]


def trimmed(min_len: int, max_len: int, too_short: str):
    """A trimmed string whose minimum-length failure carries a friendly message."""

    def check(v: str) -> str:
        v = v.strip()
        if len(v) < min_len:
            raise PydanticCustomError("too_small", too_short)
        if len(v) > max_len:
            raise PydanticCustomError("too_big", f"String must contain at most {max_len} character(s)")
        return v

    return Annotated[str, AfterValidator(check)]


def scope(me: AuthContext) -> list[Any]:
    """Params for TICKET_SELECT: workspace, viewer, and whether the viewer may see private (HR) tickets."""
    return [me.workspaceId, me.employeeId, is_exec(me.role)]


def load_ticket(db: psycopg.Connection, me: AuthContext, ticket_id: str, lock: bool = False) -> dict[str, Any]:
    if lock:
        query("SELECT 1 FROM tickets WHERE id = $1 AND workspace_id = $2 FOR UPDATE", [ticket_id, me.workspaceId], db)
    row = query_one(f"{TICKET_SELECT} AND t.id = $4", [*scope(me), ticket_id], db)
    if not row:
        raise not_found("Ticket")
    return row


def lock_team(db: psycopg.Connection, workspace_id: str, team_id: str) -> tuple[dict[str, Any], int]:
    team = query_one("SELECT id, key, name FROM ticket_teams WHERE id = $1 AND workspace_id = $2 FOR UPDATE", [team_id, workspace_id], db)
    if not team:
        raise bad_request("Unknown team")
    # Holding the team row lock serialises number allocation for this team.
    seq = query_one("SELECT COALESCE(MAX(number), 0) + 1 AS next FROM tickets WHERE team_id = $1", [team_id], db)
    return team, int(seq["next"])


def assert_employee(db: psycopg.Connection, workspace_id: str, employee_id: str) -> None:
    if not query("SELECT 1 FROM employees WHERE id = $1 AND workspace_id = $2 AND status <> 'Exited'", [employee_id, workspace_id], db):
        raise bad_request("Assignee must be an active member of this workspace")


def ticket_href(ticket_id: str) -> str:
    return f"/app/tickets?id={quote(ticket_id, safe="-_.!~*'()")}"


def notify_assignee(db: psycopg.Connection, me: AuthContext, assignee_id: str, ticket: dict[str, str]) -> None:
    if assignee_id == me.employeeId:
        return
    query(
        """INSERT INTO notifications (workspace_id, recipient_id, type, title, body, href)
     SELECT $1, $2, 'system', $3, e.name || ' assigned this ticket to you.', $4 FROM employees e WHERE e.id = $5""",
        [me.workspaceId, assignee_id, f"{ticket['identifier']} · {ticket['title']}", ticket_href(ticket["id"]), me.employeeId],
        db,
    )


# ── List & read ─────────────────────────────────────────────────────
class ListFilters(BaseModel):
    team: Annotated[str, StringConstraints(strip_whitespace=True, max_length=64)] | None = None
    status: str | None = None
    assignee: Literal["me"] | None = None
    reporter: Literal["me"] | None = None
    q: Annotated[str, StringConstraints(strip_whitespace=True, max_length=120)] | None = None


class StatusList(RootModel[list[Status]]):
    pass


@router.get("/tickets")
def list_tickets(request: Request, auth: AuthContext = Depends(require_auth)):
    f = parse(ListFilters, dict(request.query_params))
    statuses = parse(StatusList, f.status.split(",")).root if f.status else None
    params: list[Any] = scope(auth)
    where: list[str] = []
    if f.team:
        params.append(f.team)
        where.append(f"(t.team_id = ${len(params)} OR tm.key = upper(${len(params)}))")
    if statuses:
        params.append(statuses)
        where.append(f"t.status = ANY (${len(params)}::text[])")
    if f.assignee:
        where.append("t.assignee_id = $2")
    if f.reporter:
        where.append("t.reporter_id = $2")
    if f.q:
        params.append("%" + re.sub(r"[\\%_]", lambda m: "\\" + m.group(0), f.q) + "%")
        n = len(params)
        where.append(f"(t.title ILIKE ${n} OR t.identifier ILIKE ${n} OR t.description ILIKE ${n})")
    rows = query(f"{TICKET_SELECT}{''.join(f' AND {w}' for w in where)} ORDER BY t.updated_at DESC LIMIT 500", params)
    return [to_ticket(r) for r in rows]


@router.get("/tickets/{id}")
def get_ticket(id: str, auth: AuthContext = Depends(require_auth)):
    row = query_one(f"{TICKET_SELECT} AND (t.id = $4 OR t.identifier = upper($4))", [*scope(auth), id])
    if not row:
        raise not_found("Ticket")
    return to_ticket(row)


# ── Create ──────────────────────────────────────────────────────────
class CreateBody(BaseModel):
    model_config = ConfigDict(strict=True)
    teamId: Annotated[str, Field(min_length=1)]
    title: trimmed(3, 200, "Give the ticket a short title")
    description: Annotated[str, StringConstraints(strip_whitespace=True, max_length=10_000)] = ""
    status: Status = "Todo"
    priority: Priority = "None"
    assigneeId: Annotated[str, Field(min_length=1)] | None = None
    labels: Labels = []
    dueDate: Annotated[str, Field(pattern=ISO_DATE)] | None = None


@router.post("/tickets")
def create_ticket(request: Request, body: Any = Body(default=None), auth: AuthContext = Depends(require_auth)):
    b = parse(CreateBody, body)
    with tx() as db:
        team, number = lock_team(db, auth.workspaceId, b.teamId)
        if b.assigneeId:
            assert_employee(db, auth.workspaceId, b.assigneeId)
        identifier = f"{team['key']}-{number}"
        row = query_one(
            """INSERT INTO tickets (workspace_id, team_id, number, identifier, title, description, status, priority, reporter_id, assignee_id, labels, due_date, completed_at)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11::text[],$12, CASE WHEN $7 = 'Done' THEN now() END) RETURNING id""",
            [auth.workspaceId, team["id"], number, identifier, b.title, b.description, b.status, b.priority, auth.employeeId, b.assigneeId, b.labels, b.dueDate],
            db,
        )
        ticket_id = row["id"]
        if b.assigneeId:
            notify_assignee(db, auth, b.assigneeId, {"id": ticket_id, "identifier": identifier, "title": b.title})
        audit(request, auth, "ticket.created", "ticket", ticket_id, {"identifier": identifier, "team": team["key"], "priority": b.priority}, db)
        ticket = to_ticket(load_ticket(db, auth, ticket_id))
    return JSONResponse(ticket, status_code=201)


# ── Update ──────────────────────────────────────────────────────────
# Every field is optional; only assigneeId and dueDate may be null (zod `.partial().strict()`).
class PatchBody(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")
    title: trimmed(3, 200, "String must contain at least 3 character(s)") = None  # type: ignore[assignment]
    description: Annotated[str, StringConstraints(strip_whitespace=True, max_length=10_000)] = None  # type: ignore[assignment]
    status: Status = None  # type: ignore[assignment]
    priority: Priority = None  # type: ignore[assignment]
    assigneeId: Annotated[str, Field(min_length=1)] | None = None
    labels: Labels = None  # type: ignore[assignment]
    dueDate: Annotated[str, Field(pattern=ISO_DATE)] | None = None
    teamId: Annotated[str, Field(min_length=1)] = None  # type: ignore[assignment]


@router.patch("/tickets/{id}")
def patch_ticket(id: str, request: Request, body: Any = Body(default=None), auth: AuthContext = Depends(require_auth)):
    b = parse(PatchBody, body)
    given = b.model_dump(exclude_unset=True)
    if not given:
        raise bad_request("Nothing to update")

    with tx() as db:
        cur = load_ticket(db, auth, id, lock=True)
        involved = cur["reporter_id"] == auth.employeeId or cur["assignee_id"] == auth.employeeId
        # Anyone may pick up an unassigned ticket for themselves (Linear's "assign to me").
        self_assign_only = len(given) == 1 and given.get("assigneeId") == auth.employeeId and not cur["assignee_id"]
        if not involved and not is_leader(auth.role) and not self_assign_only:
            raise forbidden("Only the reporter, the assignee, managers and HR can edit this ticket")

        if given.get("assigneeId"):
            assert_employee(db, auth.workspaceId, given["assigneeId"])

        patch: dict[str, Any] = dict(given)
        identifier = cur["identifier"]
        moved = bool(given.get("teamId")) and given["teamId"] != cur["team_id"]
        if moved:
            team, number = lock_team(db, auth.workspaceId, given["teamId"])
            identifier = f"{team['key']}-{number}"
            patch.update(number=number, identifier=identifier)
        else:
            patch.pop("teamId", None)
        status_changed = bool(given.get("status")) and given["status"] != cur["status"]
        if status_changed:
            patch["completedAt"] = dt.datetime.now(dt.timezone.utc) if given["status"] == "Done" else None

        upd = build_update(patch, ["title", "description", "status", "priority", "assignee_id", "labels", "due_date", "team_id", "number", "identifier", "completed_at"], 3)
        if upd:
            set_sql, set_params = upd
            query(f"UPDATE tickets SET {set_sql}, updated_at = now() WHERE id = $1 AND workspace_id = $2", [cur["id"], auth.workspaceId, *set_params], db)

        if status_changed:
            audit(request, auth, "ticket.status_changed", "ticket", cur["id"], {"identifier": identifier, "from": cur["status"], "to": given["status"]}, db)
        if moved:
            audit(request, auth, "ticket.moved", "ticket", cur["id"], {"from": cur["identifier"], "to": identifier}, db)
        if given.get("assigneeId") and given["assigneeId"] != cur["assignee_id"]:
            notify_assignee(db, auth, given["assigneeId"], {"id": cur["id"], "identifier": identifier, "title": given.get("title") or cur["title"]})
            audit(request, auth, "ticket.assigned", "ticket", cur["id"], {"identifier": identifier, "assigneeId": given["assigneeId"]}, db)
        ticket = to_ticket(load_ticket(db, auth, cur["id"]))
    return ticket


# ── Comments ────────────────────────────────────────────────────────
class CommentBody(BaseModel):
    model_config = ConfigDict(strict=True)
    body: trimmed(1, 5000, "Write a comment first")


@router.post("/tickets/{id}/comments")
def add_comment(id: str, body: Any = Body(default=None), auth: AuthContext = Depends(require_auth)):
    text = parse(CommentBody, body).body
    with tx() as db:
        cur = load_ticket(db, auth, id)
        comment = query_one(
            "INSERT INTO ticket_comments (workspace_id, ticket_id, author_id, body) VALUES ($1, $2, $3, $4) RETURNING id, author_id, body, created_at",
            [auth.workspaceId, cur["id"], auth.employeeId, text],
            db,
        )
        query("UPDATE tickets SET updated_at = now() WHERE id = $1", [cur["id"]], db)
        # Let the other people on the ticket know (one statement, no fan-out).
        query(
            """INSERT INTO notifications (workspace_id, recipient_id, type, title, body, href)
       SELECT $1, r.id, 'system', $2, left($3, 180), $4
         FROM (SELECT DISTINCT unnest(ARRAY[$5::text, $6::text]) AS id) r
        WHERE r.id IS NOT NULL AND r.id <> $7""",
            [auth.workspaceId, f"New comment on {cur['identifier']}", text, ticket_href(cur["id"]), cur["reporter_id"], cur["assignee_id"], auth.employeeId],
            db,
        )
    return JSONResponse({"id": comment["id"], "authorId": comment["author_id"], "body": comment["body"], "createdAt": iso(comment["created_at"])}, status_code=201)
